runtime/include/mem/jemalloc/chpl-mem-impl.h

```diff
@@ -35,57 +35,50 @@
 #define CHPL_JE_TOK_PASTE(a, b) CHPL_JE_TOK_PASTE_INTERNAL(a,b)
 #define CHPL_JE_(NAME) CHPL_JE_TOK_PASTE(CHPL_JEMALLOC_PREFIX, NAME)
 
-#define CHPL_JE_CALLOC CHPL_JE_(calloc)
-#define CHPL_JE_MALLOC CHPL_JE_(malloc)
-#define CHPL_JE_POSIX_MEMALIGN CHPL_JE_(posix_memalign)
-#define CHPL_JE_REALLOC CHPL_JE_(realloc)
-#define CHPL_JE_FREE CHPL_JE_(free)
+#define CHPL_JE_MALLOCX CHPL_JE_(mallocx)
+#define CHPL_JE_RALLOCX CHPL_JE_(rallocx)
+#define CHPL_JE_DALLOCX CHPL_JE_(dallocx)
 #define CHPL_JE_NALLOCX CHPL_JE_(nallocx)
 #define CHPL_JE_MALLCTL CHPL_JE_(mallctl)
 
 
 #define MALLOCX_NO_FLAGS 0
-
-#ifdef __cplusplus
-extern "C" {
-#endif
-extern unsigned get_num_arenas(void);
-extern unsigned set_arena(unsigned);
-#ifdef __cplusplus
+// Determine which arena to use. For large allocations (32 MiB) use a dedicated
+// arena to reduce fragmentation
+extern unsigned CHPL_JE_LG_ARENA;
+static inline int CHPL_JE_MALLOCX_ARENA_FLAG(size_t size) {
+  if (size >= ((size_t) 32 << 20)) {
+    return MALLOCX_ARENA(CHPL_JE_LG_ARENA);
+  }
+  return MALLOCX_NO_FLAGS;
 }
-#endif
+
 
 static inline void* chpl_calloc(size_t n, size_t size) {
-  return CHPL_JE_CALLOC(n,size);
+  size_t nSize = n*size;
+  if (nSize == 0) { nSize = 1; }
+  return CHPL_JE_MALLOCX(nSize, MALLOCX_ZERO | CHPL_JE_MALLOCX_ARENA_FLAG(nSize));
 }
 
 static inline void* chpl_malloc(size_t size) {
-  void* ret;
-  // To limit fragmentation for large allocations use a dedicated arena
-  if (size >= (32*1024*1024)) {
-    unsigned old_arena = set_arena(get_num_arenas()-1);
-    ret = CHPL_JE_MALLOC(size);
-    set_arena(old_arena);
-  } else {
-    ret = CHPL_JE_MALLOC(size);
-  }
-  return ret;
+  if (size == 0) { size = 1; }
+  return CHPL_JE_MALLOCX(size, CHPL_JE_MALLOCX_ARENA_FLAG(size));
 }
 
 static inline void* chpl_memalign(size_t boundary, size_t size) {
-  void* ret = NULL;
-  int rc;
-  rc = CHPL_JE_POSIX_MEMALIGN(&ret, boundary, size);
-  if( rc == 0 ) return ret;
-  else return NULL;
+  if (size == 0) { size = 1; }
+  return CHPL_JE_MALLOCX(size, MALLOCX_ALIGN(boundary) | CHPL_JE_MALLOCX_ARENA_FLAG(size));
 }
 
 static inline void* chpl_realloc(void* ptr, size_t size) {
-  return CHPL_JE_REALLOC(ptr, size);
+  if (ptr == NULL) { return CHPL_JE_MALLOCX(size, CHPL_JE_MALLOCX_ARENA_FLAG(size)); }
+  if (size == 0) { CHPL_JE_DALLOCX(ptr, MALLOCX_NO_FLAGS); return NULL; }
+  return CHPL_JE_RALLOCX(ptr, size, CHPL_JE_MALLOCX_ARENA_FLAG(size));
 }
 
 static inline void chpl_free(void* ptr) {
-  CHPL_JE_FREE(ptr);
+  if (ptr == NULL) { return; }
+  CHPL_JE_DALLOCX(ptr, MALLOCX_NO_FLAGS);
```

Comment:
`DALLOC()` doesn't need the arena flag even if the pointer is to the "large" arena instead of the thread's default one?  (I guess it must not, if the tests passed ...)

Reply:
Correct, and even more so dallocx will ignore any arena flag passed to it anyways -- https://github.com/jemalloc/jemalloc/blob/e6cb7a1c9b31de3c6eca367d9164a1896bbb60ae/ChangeLog#L940

Reply:
That's a relief!